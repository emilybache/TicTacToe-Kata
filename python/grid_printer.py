

def print_grid(grid):
    """Prints any grid-like object with content at different x, y co-ordinates.

    The grid object must have:
    * attribute 'width'
    * attribute 'height'.
    * 'content_at(x, y)' which returns the single character to be printed at that location with co-ordinates (x, y).
    * attribute 'comment' (which may be empty). This is printed below the grid.
    """
    field_width = calculate_field_width(grid)
    x_coords = [_fixed_width(" ", field_width)] + [_fixed_width(i, field_width) for i in range(grid.width)]
    all_content = [x_coords]
    for j in range(grid.height):
        y_coords = [_fixed_width(j, field_width)]
        row_contents = y_coords + [_fixed_width(grid.content_at(i, j), field_width) for i in range(grid.width)]
        all_content.append(row_contents)

    all_rows = [" ".join(row) for row in all_content]
    all_rows.append("")
    grid_string = '.\n'.join(all_rows)
    return grid_string + grid.comment


def calculate_field_width(grid):
    """Given a grid with width and height, what should the fixed width of fields be for printing"""
    return max(len(str(grid.width)), len(str(grid.height)))


def _fixed_width(s, width):
    "make a string from the object s and pad with whitespace up to width"
    r = str(s)
    r += _whitespace(width - len(r))
    return r


def _whitespace(length):
    "Return a string of whitespace with the requested length"
    return " "*length

