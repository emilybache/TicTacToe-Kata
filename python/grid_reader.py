from grid import Grid
from printers import calculate_field_width


def read_grid(position):
    grid = create_grid(position)
    field_width = calculate_field_width(grid) + 1 # account for 1 padding in each field

    header_row = True
    for j, line in enumerate(position.splitlines()):
        if header_row:
            header_row = False
            continue
        line = line.replace(".", "")
        fields_and_label = _chunkstring(line, field_width)
        label = True
        for i, field in enumerate(fields_and_label):
            if label:
                label = False
                continue
            piece = field.strip()
            if piece:
                grid.content[(i-1, j-1)] = piece

    return grid


def create_grid(position):
    contents = []
    for line in position.splitlines():
        line = line.replace(".", "")
        line = line.strip()
        fields = line.split()
        if fields:
            contents.append(fields)
    width = max([len(row) for row in contents])  # header row should be longest
    height = len(contents) - 1  # ignore header row when counting height
    grid = Grid(width, height)
    return grid


def _chunkstring(string, length):
    return (string[0+i:length+i] for i in range(0, len(string), length))
