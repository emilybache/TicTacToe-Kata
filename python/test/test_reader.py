from approvaltests import verify

from grid_reader import read_grid
from grid_printer import print_grid


def test_read_empty_board():
    position = """\
  0 1 2.
0      .
1      .
2      .
"""
    grid = read_grid(position)
    assert grid.width == 3
    assert grid.height == 3


def test_read_pieces():
    position = """\
  0 1 2.
0 x    .
1   o  .
2 x    .
"""
    grid = read_grid(position)
    verify(print_grid(grid))


def test_read_larger_grid():
    position = """\
   0  1  2  3  4  5  6  7  8  9  10 11.
0  x     o                            .
1                                     .
2                                     .
3                                     .
4                                     .
5                                     .
6                                     .
7                                     .
8                                     .
9                                     .
10                         x          .
11                                    .
12                                    .
13                                    .
"""
    grid = read_grid(position)
    verify(print_grid(grid))
