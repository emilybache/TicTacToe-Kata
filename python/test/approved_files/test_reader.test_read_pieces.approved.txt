  0 1 2.
0 x    .
1   o  .
2 x    .
